Mathlib/Data/Set/Lattice.lean

```diff
@@ -0,0 +1,2522 @@
+
+-- classical -- complete_boolean_algebra
+theorem unionᵢ_eq_compl_interᵢ_compl (s : ι → Set β) : (⋃ i, s i) = (⋂ i, s iᶜ)ᶜ := by
+  simp only [compl_interᵢ, compl_compl]
+#align set.Union_eq_compl_Inter_compl Set.unionᵢ_eq_compl_interᵢ_compl
+
+-- classical -- complete_boolean_algebra
+theorem interᵢ_eq_compl_unionᵢ_compl (s : ι → Set β) : (⋂ i, s i) = (⋃ i, s iᶜ)ᶜ := by
+  simp only [compl_unionᵢ, compl_compl]
+#align set.Inter_eq_compl_Union_compl Set.interᵢ_eq_compl_unionᵢ_compl
+
+theorem inter_unionᵢ (s : Set β) (t : ι → Set β) : (s ∩ ⋃ i, t i) = ⋃ i, s ∩ t i :=
+  inf_supᵢ_eq _ _
+#align set.inter_Union Set.inter_unionᵢ
+
+theorem unionᵢ_inter (s : Set β) (t : ι → Set β) : (⋃ i, t i) ∩ s = ⋃ i, t i ∩ s :=
+  supᵢ_inf_eq _ _
+#align set.Union_inter Set.unionᵢ_inter
+
+theorem unionᵢ_union_distrib (s : ι → Set β) (t : ι → Set β) :
+    (⋃ i, s i ∪ t i) = (⋃ i, s i) ∪ ⋃ i, t i :=
+  supᵢ_sup_eq
+#align set.Union_union_distrib Set.unionᵢ_union_distrib
+
+theorem interᵢ_inter_distrib (s : ι → Set β) (t : ι → Set β) :
+    (⋂ i, s i ∩ t i) = (⋂ i, s i) ∩ ⋂ i, t i :=
+  infᵢ_inf_eq
+#align set.Inter_inter_distrib Set.interᵢ_inter_distrib
+
+theorem union_unionᵢ [Nonempty ι] (s : Set β) (t : ι → Set β) : (s ∪ ⋃ i, t i) = ⋃ i, s ∪ t i :=
+  sup_supᵢ
+#align set.union_Union Set.union_unionᵢ
+
+theorem unionᵢ_union [Nonempty ι] (s : Set β) (t : ι → Set β) : (⋃ i, t i) ∪ s = ⋃ i, t i ∪ s :=
+  supᵢ_sup
+#align set.Union_union Set.unionᵢ_union
+
+theorem inter_interᵢ [Nonempty ι] (s : Set β) (t : ι → Set β) : (s ∩ ⋂ i, t i) = ⋂ i, s ∩ t i :=
+  inf_infᵢ
+#align set.inter_Inter Set.inter_interᵢ
+
+theorem interᵢ_inter [Nonempty ι] (s : Set β) (t : ι → Set β) : (⋂ i, t i) ∩ s = ⋂ i, t i ∩ s :=
+  infᵢ_inf
+#align set.Inter_inter Set.interᵢ_inter
+
+-- classical
+theorem union_interᵢ (s : Set β) (t : ι → Set β) : (s ∪ ⋂ i, t i) = ⋂ i, s ∪ t i :=
+  sup_infᵢ_eq _ _
+#align set.union_Inter Set.union_interᵢ
+
+theorem interᵢ_union (s : ι → Set β) (t : Set β) : (⋂ i, s i) ∪ t = ⋂ i, s i ∪ t :=
+  infᵢ_sup_eq _ _
+#align set.Inter_union Set.interᵢ_union
+
+theorem unionᵢ_diff (s : Set β) (t : ι → Set β) : (⋃ i, t i) \ s = ⋃ i, t i \ s :=
+  unionᵢ_inter _ _
+#align set.Union_diff Set.unionᵢ_diff
+
+theorem diff_unionᵢ [Nonempty ι] (s : Set β) (t : ι → Set β) : (s \ ⋃ i, t i) = ⋂ i, s \ t i := by
+  rw [diff_eq, compl_unionᵢ, inter_interᵢ]; rfl
+#align set.diff_Union Set.diff_unionᵢ
+
+theorem diff_interᵢ (s : Set β) (t : ι → Set β) : (s \ ⋂ i, t i) = ⋃ i, s \ t i := by
+  rw [diff_eq, compl_interᵢ, inter_unionᵢ]; rfl
+#align set.diff_Inter Set.diff_interᵢ
+
+theorem directed_on_unionᵢ {r} {f : ι → Set α} (hd : Directed (· ⊆ ·) f)
+    (h : ∀ x, DirectedOn r (f x)) : DirectedOn r (⋃ x, f x) := by
+  simp only [DirectedOn, exists_prop, mem_unionᵢ, exists_imp];
+    exact fun a₁ b₁ fb₁ a₂ b₂ fb₂ =>
+      let ⟨z, zb₁, zb₂⟩ := hd b₁ b₂
+      let ⟨x, xf, xa₁, xa₂⟩ := h z a₁ (zb₁ fb₁) a₂ (zb₂ fb₂)
+      ⟨x, ⟨z, xf⟩, xa₁, xa₂⟩
```

Comment:
```suggestion
  simp only [DirectedOn, exists_prop, mem_unionᵢ, exists_imp]
  exact fun a₁ b₁ fb₁ a₂ b₂ fb₂ =>
    let ⟨z, zb₁, zb₂⟩ := hd b₁ b₂
    let ⟨x, xf, xa₁, xa₂⟩ := h z a₁ (zb₁ fb₁) a₂ (zb₂ fb₂)
    ⟨x, ⟨z, xf⟩, xa₁, xa₂⟩
```
Not sure why that semicolon was there? Autoporter suggested <;> but there's only one goal.